tests/test_qasm_examples.py

```diff
@@ -1,17 +1,136 @@
 import os
 import re
+import sympy
 import pytest
+import numpy as np
 
 from qdata import parser
 
+
+def split_at_semicol(qasm_lines):
+    """Helper function to split lines at semicolons"""
+    form_lines = []
+    for line_1 in qasm_lines:
+        split_lines = line_1.split(";")
+        if len(split_lines) == 1:
+            form_lines.append(split_lines[0])
+        elif len(split_lines) == 2 and split_lines[1] == "":
+            form_lines.append(split_lines[0] + ";")
+        else:
+            for line_2 in split_lines:
+                delim = ";" * (clean(line_2) not in ["{", "}"])
+                if clean(line_2):
+                    form_lines.append(clean(line_2) + delim)
+    return form_lines
+
+
+def fix_braces(qasm_lines):
+    """Helper function to place braces on separate lines"""
+    form_lines = []
+    for line_1 in qasm_lines:
+        # if a line contains empty braces, format them separately
+        match = re.search(r"{\s*}", line_1)
+        if match:
+            form_lines.append(line_1.replace(match.group(), "{ }"))
+            continue
+
+        # split at "{" and re-add them as separate list-elements
+        res = line_1.split("{")
+        next_qasm_lines = [clean(em) for el in list(zip(["{"]*len(res), res)) for em in el if clean(em)][1:]
+        for line_2 in next_qasm_lines:
+            # split the new list at "}" and re-add them as separate list-elements
+            res = line_2.split("}")
+            lines = [clean(em) for el in list(zip(["}"]*len(res), res)) for em in el if clean(em)][1:]
+            form_lines.extend(lines)
+
+    #asdad
+    tab = False
+    for i in range(1, len(form_lines) - 1):
+        if tab and form_lines[i] != "}":
+            form_lines[i] = "    " + form_lines[i]
+
+        if form_lines[i] == "{":
+            tab = True
+        elif form_lines[i] == "}":
+            tab = False
+
+    return form_lines
+
+
+def fix_ifs(line):
+    """Helper function to format if-statements correctly"""
+    p = re.compile(r"(if)\s*\(\s*(\w+)\s*(==|!=|>=|<=|<|>)\s*(\w+)\s*\)")
+    match = p.match(line)
+    # if an if-string is found, "build up" a correctly formatted string
+    if match:
+        ifstring = match.group(1) + " (" + match.group(2) + " " + match.group(3) + " " + match.group(4) + ")"
+        return ifstring + line[match.span()[1]:]
+    return line
+
+
+def fix_spaces(line):
+    """Helper function to format spaces before/after paranthesis and after commas correctly"""
+    line = ",".join(re.split(r"\s*,\s*", line))
+    line = "(".join(re.split(r"\s*\(\s*", line))
+    line = ") ".join(re.split(r"\s*\)\s*", line))
+    return clean(line)
```

Comment:
Was this @co9olguy original space normalizing logic? Or is this new?

Reply:
My functions would just turn any whitespace into a single space. I think this is to insert whitespace in places where it might not be explicitly?

Reply:
It inserts/removes whitespaces in front of and behind commas and parantheses so that it always looks like e.g. `u(a,b,c)` and _never_ `u (a,b, c )`. We might want to change these rules, although what they do now is making sure it's formatted the same way as they're currently output.